src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        return findByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexValidation(index);
+        currentNode = findByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        return unlink(findByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        currentNode = first;
+        if (object == null) {
+            while (currentNode != null) {
+                if (currentNode.item == null) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        } else {
+            while (currentNode != null) {
+                if (object.equals(currentNode.item)) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return currentSize;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return currentSize == 0;
+    }
+
+    private void indexValidation(int index) {
+        if (index < 0 || index >= currentSize) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void linkFirst(T value) {
+        if (currentSize == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+        }
+        currentSize++;
+    }
+
+    private void linkLast(T value) {
+        newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        last = newNode;
+        currentSize++;
+    }
+
+    private void linkMiddle(T value, int index) {
+        indexValidation(index);
+        currentNode = findByIndex(index);
+        newNode = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        currentSize++;
+    }
+
+    private T unlink(Node<T> node) {
+        T value = node.item;
+        if (node.prev == null && node.next != null) {
+            node.next.prev = null;
+            first = node.next;
+        } else if (node.next == null && node.prev != null) {
+            node.prev.next = null;
+        } else if (node.next != null) {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        currentSize--;
+        return value;
+    }
```

Comment:
what if `node.prev == null && node.next == null`? (case when we have only one node)